### PR TITLE
add applicative cache on process definition service to improve perfor…

### PR DESCRIPTION
**Pull request by baptistemesta**
…mance

### CODE REVIEW

File: bpm/bonita-core/bonita-process-definition/bonita-process-definition-impl/src/main/java/org/bonitasoft/engine/core/process/definition/ProcessDefinitionServiceImpl.java [hunk 1]
```diff
@@ -176,14 +182,22 @@ public void disableProcessDeploymentInfo(final long processId) throws SProcessDe
...
             log(processDefinitionDeployInfo.getId(), SQueriableLog.STATUS_FAIL, logBuilder, "disableProcess");
             throw new SProcessDisablementException(e);
         }
     }
 
+    void updateLastUpdateDateInCache(long processId, SProcessDefinitionDeployInfo processDefinitionDeployInfo) throws SCacheException {
+        final Pair<Long, SProcessDefinition> fromCache = getFromCache(processId);
+        if (fromCache != null) {
+            fromCache.setKey(processDefinitionDeployInfo.getLastUpdateDate());
+            cacheService.store(PROCESS_CACHE_NAME, processId, fromCache);
```
Comment (educhastenier) [inline bpm/bonita-core/bonita-process-definition/bonita-process-definition-impl/src/main/java/org/bonitasoft/engine/core/process/definition/ProcessDefinitionServiceImpl.java]: if getFromCache() != null, there is no need to store it again, as it is already in the cache service

File: bpm/bonita-core/bonita-process-definition/bonita-process-definition-impl/src/main/java/org/bonitasoft/engine/core/process/definition/ProcessDefinitionServiceImpl.java [hunk 2]
```diff
@@ -254,15 +268,28 @@ public long getNumberOfProcessDeploymentInfos() throws SProcessDefinitionReadExc
     public SProcessDefinition getProcessDefinition(final long processId) throws SProcessDefinitionNotFoundException, SProcessDefinitionReadException {
         try {
             final SProcessDefinitionDeployInfo processDeploymentInfo = getProcessDeploymentInfo(processId);
```
Comment (educhastenier) [inline bpm/bonita-core/bonita-process-definition/bonita-process-definition-impl/src/main/java/org/bonitasoft/engine/core/process/definition/ProcessDefinitionServiceImpl.java]: // Read from database (or Hibernate cache):

File: bpm/bonita-core/bonita-process-definition/bonita-process-definition-impl/src/main/java/org/bonitasoft/engine/core/process/definition/ProcessDefinitionServiceImpl.java [hunk 3]
```diff
@@ -254,15 +268,28 @@ public long getNumberOfProcessDeploymentInfos() throws SProcessDefinitionReadExc
...
-            final DesignProcessDefinition objectFromXML = processDefinitionBARContribution.convertXmlToProcess(processDeploymentInfo.getDesignContent().getContent());
-            SProcessDefinition sProcessDefinition = convertDesignProcessDefinition(objectFromXML);
-            setIdOnProcessDefinition(sProcessDefinition, processId);
+            final Pair<Long, SProcessDefinition> processWithTimestamp = getFromCache(processId);
```
Comment (educhastenier) [inline bpm/bonita-core/bonita-process-definition/bonita-process-definition-impl/src/main/java/org/bonitasoft/engine/core/process/definition/ProcessDefinitionServiceImpl.java]: // Read it from the cache:

File: bpm/bonita-core/bonita-process-definition/bonita-process-definition-impl/src/main/java/org/bonitasoft/engine/core/process/definition/ProcessDefinitionServiceImpl.java [hunk 4]
```diff
@@ -254,15 +268,28 @@ public long getNumberOfProcessDeploymentInfos() throws SProcessDefinitionReadExc
...
+            SProcessDefinition sProcessDefinition;
+            if (processWithTimestamp == null || processWithTimestamp.getKey() != processDeploymentInfo.getLastUpdateDate()) {
```
Comment (educhastenier) [inline bpm/bonita-core/bonita-process-definition/bonita-process-definition-impl/src/main/java/org/bonitasoft/engine/core/process/definition/ProcessDefinitionServiceImpl.java]: // ... and compare both to check if cache is up-to-date:

File: bpm/bonita-core/bonita-process-definition/bonita-process-definition-impl/src/main/java/org/bonitasoft/engine/core/process/definition/ProcessDefinitionServiceImpl.java (hunk 1 above)
Comment (baptistemesta) [inline bpm/bonita-core/bonita-process-definition/bonita-process-definition-impl/src/main/java/org/bonitasoft/engine/core/process/definition/ProcessDefinitionServiceImpl.java]: I update here the last update date in cache, I call store to store the updated version.
If i do not call the store the value will not be updated in cache if the cache configuration have copy on read or copy on write activated

Comment (educhastenier) [inline bpm/bonita-core/bonita-process-definition/bonita-process-definition-impl/src/main/java/org/bonitasoft/engine/core/process/definition/ProcessDefinitionServiceImpl.java]: Ok, got it.
On another hand, **getFromCache()** is a wrapping method, and **cacheService.store()** is lower-level method called directly, it seems not homogeneous.

Comment (baptistemesta) [inline bpm/bonita-core/bonita-process-definition/bonita-process-definition-impl/src/main/java/org/bonitasoft/engine/core/process/definition/ProcessDefinitionServiceImpl.java]: you're right, i'll change that

File: bpm/bonita-core/bonita-process-definition/bonita-process-definition-impl/src/test/java/org/bonitasoft/engine/core/process/definition/ProcessDefinitionServiceImplTest.java [hunk 5]
```diff
@@ -360,6 +390,43 @@ public void getProcessDefinitionId_should_throw_SProcessDefinitionReadException_
...
+    public void getProcessDefinition_from_database_when_cache_is_outdated() throws Exception {
+        sProcessDefinitionDeployInfo.setLastUpdateDate(5478L);
+        final SProcessDefinitionImpl processDefinitionB = new SProcessDefinitionImpl("a", "b");
+        final SProcessDefinitionImpl processDefinitionC = new SProcessDefinitionImpl("a", "c");
+        doReturn(new Pair<Long, SProcessDefinition>(5477L, processDefinitionB)).when(processDefinitionServiceImpl).getSProcessDefinitionFromCache(PROCESS_ID);
+        doReturn(processDefinitionC).when(processDefinitionServiceImpl).convertDesignProcessDefinition(designProcessDefinition);
+        final SProcessDefinition processDefinition = processDefinitionServiceImpl.getProcessDefinition(PROCESS_ID);
+
+        assertThat(processDefinition).as("returned process definition from cache").isEqualTo(processDefinitionC);
+        verify(cacheService, times(1)).store(anyString(), any(Serializable.class), any());
```
Comment (educhastenier) [inline bpm/bonita-core/bonita-process-definition/bonita-process-definition-impl/src/test/java/org/bonitasoft/engine/core/process/definition/ProcessDefinitionServiceImplTest.java]: Should you check that the call to the **store()** method is done with the right argument **processDefinitionC** ?

**Comment by educhastenier** [comment]
Ok to merge once CP 5 has been merged to **master** already

**Pull request merged**